### PR TITLE
Switch out spotless for an alternative

### PR DESCRIPTION
**Pull request by lukebemish**
Spotless has... some quirks. Among others:
- the author dislikes gradle's configuration cache conceptually, and spotless's support for it is generally iffy, including using some... questionable workarounds
- a good bit of formatting logic is dependent on the JVM used to invoke _gradle_, and cannot change with the toolchain -- this is... not optimal, when you'd like a neo-dev environment to be pretty easy to have Just Work on most machines with a somewhat recent JVM present locally.
- Sometimes its cache gets... wacky, and just doesn't recognize files as needing re-formatted
- Running gradle with a different java version than you are developing for causes... issues, due to how spotless invokes formatters.

`immaculate` is a spotless alternative built with a much narrower scope (namely, the sort of formatting you tend to see in java projects like this), designed to utilize built-in gradle features when possible (up-to-date checking and incremental tasks for figuring out which files to re-check), prefer lazy configuration, and respect gradle best practices whereever I have any clue what they are; this PR switches neo-dev to use it instead of spotless. Note that, due to using the same version of eclipse JDT and the same other formatting steps besides that, no formatting has shifted compared to using spotless.

`immaculate` is new and thus rather untested; `spotless` is a much more established plugin, but also has a much more established set of issues, some of which seem to be continually popping up in discord (the notable one here is the whole bit where tons of stuff is dependent on the java version used to invoke gradle). Furthermore, this change could probably use some more extensive testing to make sure it behaves right in a number of different development environments.

In any case, this PR could use testing on different environments to make sure it's all behaving sensibly across dev environments.

### CODE REVIEW

File: buildSrc/src/main/groovy/neoforge.formatting-conventions.gradle [hunk 1]
```diff
@@ -28,15 +28,19 @@ final generatePackageInfos = tasks.register('generatePackageInfos', Task) {
...
+immaculate {
+    workflows.register('java') {
+        java()
+        trailingNewline()
+        noTabs()
+        googleFixImports()
+        toggleOff = 'spotless:off'
+        toggleOn = 'spotless:on'
+        eclipse {
+            version '3.37.0'
```
Comment (sciwhiz12) [inline buildSrc/src/main/groovy/neoforge.formatting-conventions.gradle:48]: This looks like a downgrade from 4.31 to 3.37.0. Why?

File: buildSrc/src/main/groovy/neoforge.formatting-conventions.gradle [hunk 2]
```diff
@@ -28,15 +28,19 @@ final generatePackageInfos = tasks.register('generatePackageInfos', Task) {
...
-        toggleOffOn()
-        // Pin version to 4.31 because of a Spotless bug https://github.com/diffplug/spotless/issues/1992
-        eclipse('4.31').configFile rootProject.file('codeformat/formatter-config.xml')
-        importOrder()
+immaculate {
+    workflows.register('java') {
+        java()
+        trailingNewline()
+        noTabs()
+        googleFixImports()
```
Comment (sciwhiz12) [inline buildSrc/src/main/groovy/neoforge.formatting-conventions.gradle:44]: This is the equivalent to `importOrder()` from Spotless?

File: buildSrc/src/main/groovy/neoforge.formatting-conventions.gradle [hunk 3]
```diff
@@ -55,28 +59,27 @@ spotless {
         custom 'jetbrainsNullable', { String fileContents ->
             fileContents.replace('javax.annotation.Nullable', 'org.jetbrains.annotations.Nullable')
         }
-        bumpThisNumberIfACustomStepChanges(3)
```
Comment (sciwhiz12) [inline buildSrc/src/main/groovy/neoforge.formatting-conventions.gradle:66]: Could you explain why this line was removed? Particularly, how Immaculate differs from Spotless in how it handles custom steps and their changes.

**Review by sciwhiz12 — COMMENTED** [review]
Untested yet, but looks clean enough. I have a few spots that need some explanation, though.

File: buildSrc/src/main/groovy/neoforge.formatting-conventions.gradle (hunk 3 above)
Comment (lukebemish) [reply buildSrc/src/main/groovy/neoforge.formatting-conventions.gradle:66]: All caching goes through normal up-to-date checking -- a change to the custom step implementation will make the up-to-date check fail and thus, all this is handled automatically.

File: buildSrc/src/main/groovy/neoforge.formatting-conventions.gradle (hunk 1 above)
Comment (lukebemish) [reply buildSrc/src/main/groovy/neoforge.formatting-conventions.gradle:48]: Eclipse uses different versions for different things. I forget all the proper names but version 4.31 of the one depends on 3.37.0 of the other -- suffice to say, the one currently used in neo is 3.37.0

File: buildSrc/src/main/groovy/neoforge.formatting-conventions.gradle (hunk 2 above)
Comment (lukebemish) [reply buildSrc/src/main/groovy/neoforge.formatting-conventions.gradle:44]: importOrder() and removeUnusedImports() -> googleFixImports()
indentWithSpaces() -> noTabs()
endWithNewline() -> trailingNewline()

File: buildSrc/src/main/groovy/neoforge.formatting-conventions.gradle (hunk 1 above)
Comment (lukebemish) [reply buildSrc/src/main/groovy/neoforge.formatting-conventions.gradle:48]: For reference: the newest version of `org.eclipse.jdt:org.eclipse.jdt.core` (the thing with the actual formatter) is `3.39.0`.

File: buildSrc/src/main/groovy/neoforge.formatting-conventions.gradle (hunk 3 above)
Comment (lukebemish) [reply buildSrc/src/main/groovy/neoforge.formatting-conventions.gradle:66]: Following up on this further: immaculate just uses normal old up-to-date checking _everywhere_ combined with incremental file collections to not reprocess every single file when one changes. This means, basically, that you shouldn't get the occasional messed up caches issue like spotless could have, and that you don't need any custom logic like spotless needed for custom steps to have functional up-to-date checking, as the implementation is part of the key.

**Review by Technici4n — DISMISSED** [review]

File: docs/CONTRIBUTING.md [hunk 4]
```diff
@@ -27,7 +27,7 @@ Contributing to NeoForge
    - If possible, write an automated test under the tests project. See [NEOGAMETESTS.md](NEOGAMETESTS.md) for more info.
 10. Run `gradlew genPatches` to generate patch-files from the patched sources
 11. Run `gradlew applyAllFormatting` to automatically format sources
-12. Check correct formatting with `gradlew spotlessCheck`
+12. Check correct formatting with `gradlew immaculateCheck`
```
Comment (sciwhiz12) [inline docs/CONTRIBUTING.md]: Should we perhaps recommend that `checkFormatting` task instead, as seen in the `neoforge.formatting-conventions` script plugin? It'd also check the license headers, and would be consistent with what the GitHub Actions runs for the PR checks.

Comment (lukebemish) [reply docs/CONTRIBUTING.md]: Made that change; I agree, it's more sensible.

**Review by Technici4n — APPROVED** [review]

**Comment by Matyrobbrt** [comment]
@neoforged/bots backport 1.21.1

**Comment by Matyrobbrt** [comment]
@neoforged/bots backport 1.21.1

**Pull request merged**